libpromises/evalfunction.c

```diff
@@ -6659,6 +6659,188 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
     return (FnCallResult) { FNCALL_SUCCESS, (Rval) { json, RVAL_TYPE_CONTAINER } };
 }
 
+static int ContainerComparator(JsonElement const *left_obj,
+                               JsonElement const *right_obj,
+                               void *user_data)
+{
+    size_t const *index = user_data;
+
+    char const *left = JsonPrimitiveGetAsString(JsonAt(left_obj, *index));
+    char const *right = JsonPrimitiveGetAsString(JsonAt(right_obj, *index));
+    return strcmp(left, right);
+}
+
+static FnCallResult FnCallClassFilterCsv(EvalContext *ctx,
+                                         ARG_UNUSED Policy const *policy,
+                                         FnCall const *fp,
+                                         Rlist const *args)
+{
+    if (args == NULL || args->next == NULL || args->next->next == NULL)
+    {
+        FatalError(ctx, "Function %s requires at least 3 arguments",
+                   fp->name);
+    }
+
+    char const *path = RlistScalarValue(args);
+    bool const has_heading = BooleanFromString(RlistScalarValue(args->next));
+    size_t const class_index = IntFromString(RlistScalarValue(args->next->next));
+    Rlist const *sort_arg = args->next->next->next;
+
+    FILE *csv_file = safe_fopen(path, "r");
+    if (csv_file == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Failed to read file %s: %s",
+            fp->name, path, GetErrorStrFromCode(errno));
+        FnFailure();
+    }
+
+    Seq *heading = NULL;
+    JsonElement *json = JsonArrayCreate(50);
+    char *line;
+    size_t num_columns = 0;
+
+    // Current line number, for debugging
+    size_t line_number = 0;
+
+    while ((line = GetCsvLineNext(csv_file)) != NULL)
```

Comment:
I suppose if you wanted to get real wild you could put something here like:

```
if (has_heading)
{
  heading = GetCsvLineNext(csv_file)
  // check for null, error maybe, ...
}

while ((line = GetCsvLineNext(csv_file)) != NULL)
```

So that getting the header, which is the first line, is first when reading the code.

Reply:
The reason I didn't want to do that is because I would have to duplicate all of the error-checking code.

Reply:
...though splitting it up in functions and putting this in `json-utils.c` or similar is probably an even better solution.